ads/google/utils.js

```diff
@@ -14,16 +14,280 @@
+    formatFixedWidthInteger(slotTop * 5 / viewHeight, 4) +
+    // ad slot left, in 1/5 viewport width units
+    formatFixedWidthInteger(slotLeft * 5 / viewWidth, 4);
+}
+
+/**
+ * Make the URL for an ad request, from an amp-ad element.
+ * @param {number} slotNumber
+ * @param {!Window} global
+ * @param {!Object} data
+ * @param {!IntersectionObserverEntry} intersectionRecord
+ * @return {string}
+ */
+export function adsenseRequestURLForAmpAd(slotNumber, global, data,
+                                          intersectionRecord) {
+  // Assumes that data has been checked for valid parameters, via checkData.
+  return adsenseRequestURL(global.context.startTime, slotNumber, global,
+                           data, 1, global.context.canonicalUrl,
+                           getCorrelator(global), intersectionRecord);
+}
+
+/**
+ * @private {number}
+ */
+const MAX_URL_LENGTH_ = 4096;
+
+/**
+ * @private {number}
+ */
+const MIN_USEFUL_URL_PARAM_LENGTH_ = '&url=http%3A%2F%2F'.length + 1;
+
+/**
+ * @private {string}
+ */
+const TRUNC_PARAM_ = '&trunc=1';
+
+/**
+ * Make the URL for an ad request.
+ * @param {number} startTime
+ * @param {number} slotNumber
+ * @param {!Window} global
+ * @param {!Object} data
+ * @param {number) isAmp
+ * @param {string} canonicalUrl
+ * @param {number} correlator
+ * @param {!IntersectionObserverEntry} intersectionRecord
+ * @return {string}
+ */
+function adsenseRequestURL(startTime, slotNumber, global, data, isAmp,
+                           canonicalUrl, correlator, intersectionRecord) {
+  // const lmt = global.document.lastModified;
+  const slot = intersectionRecord.boundingClientRect;
+  // For this intersectionRecord, root is the viewport.
+  const width = slot.width;
+  const height = slot.height;
+  const viewport = intersectionRecord.rootBounds;
+  const screen = global.screen;
+  // const fractionInViewport = intersectionRecord.intersectionRatio;
+  const visibilityState = documentStateFor(global).getVisibilityState();
+  const visNum = {'visible': 1,
+                  'hidden': 2,
+                  'prerender': 3,
+                  // Not returned by getVisibilityState.
+                  'preview': 4,
+                  // Not expected by ad requests.
+                  'unloaded': 0}[visibilityState] || 0;
+  const adtest = data['adtest'];
+  const adClient = data['adClient'];
+  const adSlot = data['adSlot'];
+  let experimentIds = null;
+  if (data['experimentId'] &&
+      validateExperimentIds(parseExperimentIds(data['experimentId']))) {
+    experimentIds = data['experimentId'];
+  }
+
+  let url = `https://googleads.g.doubleclick.net/pagead/ads?is_amp=${isAmp}` +
+    // Protect against wildly long client and slot ids.
+    (adClient ? `&client=${encodeURIComponent(adClient.substr(0, 50))}` : '') +
+    (adSlot ? `&slotname=${encodeURIComponent(adSlot.substr(0, 50))}` : '') +
+    '&d_imp=1&output=html' +
+    `&format=${width}x${height}&w=${width}&h=${height}` +
```

Comment:
width and height values came out of the IntersectionRecord, but maybe you don't trust that?
I'll encode them.